Youtube Trending Feed Scrapper/youtube_scrapper.py

```diff
@@ -0,0 +1,143 @@
+    title = mongoengine.StringField(required=True)
+    channel = mongoengine.StringField(required=True)
+    link = mongoengine.StringField(required=True)
+    views = mongoengine.StringField(required=True)
+    date = mongoengine.StringField(required=True)
+
+    meta = {"indexes": ["section"]}
+
+
+def load_driver():
+    """Load Chrome webdriver."""
+    driver = webdriver.Chrome("chromedriver.exe")
+    return driver
+
+
+def page_scrap(driver):
+    """Scrap YouTube trending feed."""
+    # pages to be scrapped: Now, Music, Gaming, Movies
+    pages = ["https://www.youtube.com/feed/trending",
+             "https://www.youtube.com/feed/trending?bp=4gINGgt5dG1hX2NoYXJ0cw%3D%3D",
+             "https://www.youtube.com/feed/trending?bp=4gIcGhpnYW1pbmdfY29ycHVzX21vc3RfcG9wdWxhcg%3D%3D",
+             "https://www.youtube.com/feed/trending?bp=4gIKGgh0cmFpbGVycw%3D%3D"]
+    sections = ["Now", "Music", "Gaming", "Movies"]
+
+    for num in range(4):
+        driver.get(pages[num])
+        time.sleep(3)  # Make sure that all the page is loaded
+        # Extract first 10 contents
+        cards = driver.find_elements_by_tag_name("ytd-video-renderer")[:10]
+        links = driver.find_elements_by_id("video-title")[:10]
+        meta_data = driver.find_elements_by_tag_name("ytd-video-meta-block")[:10]
+        for i in range(10):
+            # Splitted meta data that will be saved
+            meta_splitted = meta_data[i].text.split("\n")
+            # Sometimes this character is extracted for unknown reasons
+            try:
+                meta_splitted.remove("•")
+            except:
+                pass
+            section = sections[num]     # Scrapped from which section?
+            link = links[i].get_attribute("href")  # Video Link
+            title   = links[i].text     # Video title
+            channel = meta_splitted[0]  # Channel name
+            views   = meta_splitted[1]  # Video Views
+            date    = meta_splitted[2]  # Release date
+
+            """Arguments validation is better than making a scrapping algorithm for each"""
+            if mongo:
+                save_to_db(section, title, channel, link, views, date)
+            if csv:
+                append_to_df(section, title, channel, link, views, date)
+                
+        print(f"[+]Finished scraping '{sections[num]}' section!")
+
+    # last validation for csv
+    if csv:
+        save_to_csv()
+
+
+def save_to_db(section, title, channel, link, views, date):
+    """Saves a record to database."""
+    # Create object
+    record = Trending(
+                section = section,
+                title = title,
+                channel = channel,
+                link = link,
+                views = views,
+                date = date)
+    # Save record
+    record.save()
+
+
+def append_to_df(section, title, channel, link, views, date):
+    """Appends a record to dataframe."""
+    global df
+    df = df.append({"section": section,
```

Comment:
this is a very common error while deploying data science applications in the cloud. I have faced this issue many times. The solution to this define this "df" as empty data frame at the beginning of the script

Reply:
Hmmm, If I did that then an empty DataFrame will be initialized with every call of this function causing the discard of any previous data that is stored in it. :thinking:

Reply:
or do you mean defining it at the beginning of the file?

Reply:
At the beginning of the file

Reply:
hmm ok, lemme edit that.